master-server/src/main/java/com/replicated_log/master_server/controller/rest/MasterWithBrokerRestController.java

```diff
@@ -0,0 +1,121 @@
+package com.replicated_log.master_server.controller.rest;
+
+import com.replicated_log.master_server.model.Ack;
+import com.replicated_log.master_server.service.AckService;
+import com.replicated_log.master_server.model.AckStatusCode;
+import com.replicated_log.master_server.model.Address;
+import com.replicated_log.master_server.service.AddressService;
+import com.replicated_log.master_server.model.HealthStatus;
+import com.replicated_log.master_server.service.BrokerService;
+import com.replicated_log.master_server.model.Item;
+import com.replicated_log.master_server.service.ItemService;
+import com.replicated_log.master_server.service.PingService;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Set;
+
+import static org.springframework.http.HttpStatus.SERVICE_UNAVAILABLE;
+
+@RestController
+@RequestMapping(MasterWithBrokerRestController.MASTER_URL)
+@CrossOrigin("*")
+public class MasterWithBrokerRestController {
+
+    private final Logger LOG = LogManager.getLogger(MasterWithBrokerRestController.class);
+
+    public static final String MASTER_URL = "/master";
+
+    @Value("${master.server.baseurl}")
+    private String MASTER_SERVER_BASE_URL;
+
+    @Autowired
+    private ItemService itemService;
+
+    @Autowired
+    private AddressService addressService;
+
+    @Autowired
+    private AckService ackService;
+
+    @Autowired
+    private BrokerService brokerService;
+
+    @Autowired
+    private PingService pingService;
+
+
+    @PostMapping("/item/{w}")
+    public ResponseEntity appendItem(@RequestBody Item item, @PathVariable int w) {
+        LOG.info("С--> Start appendItem POST, w = " + w);
+
+        if (!pingService.hasQuorum()) {
+            return ResponseEntity.status(SERVICE_UNAVAILABLE).body("There is no quorum. Master switched into read-only mode and doesn`t accept messages append requests");
+        }
+
+        item.setW(w);
+
+        Item itemToSend = itemService.addItem(item);
+        LOG.info("С--> Added to Master: " + itemToSend);
+        LOG.info("С--> Set of Master:" + itemService.getItems());
+
+        ackService.addItemToItemAcksMap(itemToSend);
+        //because one ack is from Master
+        ackService.addAckForItem(itemToSend.getId(), new Ack(MASTER_SERVER_BASE_URL, itemToSend.getId(), AckStatusCode.SUCCESS));
+
+        brokerService.addItemToBrokerMap(itemToSend);
+
+        while ((int) ackService.getNumberOfAcksOfItem(itemToSend.getId()) < itemToSend.getW()) {
```

Comment:
as I understand
 in this endpoint there is  the functionality of sending the item,
please correct me if I am mistaken, it is blocked operation 

until the all acks are received by master node, the new item can not be sent?

Reply:
correct, we are blocked until we receive all necessary acks from Secondaries for the item. And if we start new Rest client that is running in parallel it won`t be blocked according to Project Requirements.